OCPQE-27354: Add the CI job to cover the migration from a public cluster to private on Azure

jinyunma opened the pull request:
Add two jobs based on doc https://docs.openshift.com/container-platform/4.17/post_installation_configuration/configuring-private-cluster.html#configuring-private-cluster:
- azure-ipi-sharednetwork-public-to-private
- azure-ipi-public-to-private

currently, only support 4.17+ where installer uses CAPI method.
In terraform way, need step to create outbound rule in external lb.

jinyunma commented:
/pj-rehearse periodic-ci-openshift-verification-tests-master-installer-rehearse-4.18-installer-rehearse-azure

jinyunma commented:
/pj-rehearse periodic-ci-openshift-verification-tests-master-installer-rehearse-4.18-installer-rehearse-azure

gpei commented on ci-operator/step-registry/azure/cluster/public-to-private/azure-cluster-public-to-private-commands.sh:
```diff
@@ -0,0 +1,204 @@
+# az should already be there
+command -v az
+az --version
+
```
@jinyunma if the current step only works for 4.17+, how about adding an OCP version check here, if it was added to 4.16/4.15 jobs by accident, it can quit directly.

jinyunma replied:
sure, updated.

jinyunma commented:
/pj-rehearse periodic-ci-openshift-verification-tests-master-installer-rehearse-4.18-installer-rehearse-azure

gpei commented on ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.17__multi-nightly.yaml:
```diff
@@ -1019,6 +1019,46 @@ tests:
     test:
     - chain: openshift-e2e-test-qe
     workflow: cucushift-installer-rehearse-azure-ipi-proxy
+- as: azure-ipi-public-to-private-amd-f28-destructive
```
My thought was, if we turned the public cluster into a private one, and passed the private cluster check, it supposed to be a private cluster just like a fresh installed one, and we've got many private CI jobs with E2E, so I just added it to the installation test yaml.

jinyunma replied:
Because this change also impact image registry, so add jobs in e2e test.

gpei replied:
Let me move the AWS public-to-private jobs to nightly to keep consistent with Azure later.

gpei replied:
Updated in https://github.com/openshift/release/pull/59604

gpei commented on ci-operator/step-registry/azure/provision/bastionhost/azure-provision-bastionhost-commands.sh:
```diff
@@ -181,20 +187,21 @@ vhd_blob_url=$(az storage blob url --account-name ${sa_name} --account-key ${acc
+    run_command "az network nsg create -g ${bastion_rg} -n ${bastion_nsg}" &&
+    run_command "az network nsg rule create -g ${bastion_rg} --nsg-name '${bastion_nsg}' -n '${bastion_name}-allow' --priority 1000 --access Allow --source-port-ranges '*' --destination-port-ranges ${open_port}" &&
+    #subnet cidr for int service is hard code, it should be a sub rang of the whole VNet cidr, and not conflicts with master subnet and worker subnet
+    bastion_subnet_cidr="10.0.99.0/24"
```
I see we introduced the new env var `BASTION_SUBNET_CIDR` with default: "10.0.99.0/24", but it's not being used in this script.

jinyunma replied:
yes, should be removed.

jianlinliu commented on ci-operator/step-registry/azure/provision/bastionhost/azure-provision-bastionhost-commands.sh:
```diff
@@ -92,6 +93,11 @@ if [ -z "${VNET_NAME}" ]; then
     echo "Did not find ${vnet_file}!"
     exit 1
   fi
+elif [[ "${VNET_NAME}" == "PRESERVED_FROM_CLUSTER" ]]; then
```
I guess we do not have to use `PRESERVED_FROM_CLUSTER` here, just judge if `${SHARED_DIR}/metadata.json` exits or not, if it exists, we can set vnet name and subnet for bastion host. 

The priority would be like `VNET_NAME` > `metadata.json`

jianlinliu commented on ci-operator/step-registry/azure/provision/bastionhost/azure-provision-bastionhost-commands.sh:
```diff
@@ -84,6 +84,7 @@ else
   bastion_rg="${RESOURCE_GROUP}"
```
Maybe we do not have to call `azure-provision-resourcegroup` step in this case. We can run a common IPI install, after that, we can get resource group info from metadata.json.

The priority would be like RESOURCE_GROUP > ${SHARED_DIR}/resourcegroup > metadata.json

And maybe we can exit 0 if did not find any resource group info after the above query search.

jinyunma replied:
okay, and there might have product bug when creating cluster in existing resource group.
will change to use "RESOURCE_GROUP > ${SHARED_DIR}/resourcegroup > metadata.json".

jianlinliu commented on ci-operator/step-registry/azure/provision/bastionhost/azure-provision-bastionhost-ref.yaml:
```diff
@@ -22,7 +22,11 @@ ref:
     documentation: |-
-      The vnet name that is already existing.
+      The vnet name that is already existing, or "PRESERVED_FROM_CLUSTER" to use cluster vnet.
+  - name: BASTION_SUBNET_CIDR
```
Seem like we did not reference this ENV var in scripts.

jinyunma replied:
yes, should be removed.

jianlinliu commented on ci-operator/step-registry/azure/cluster/public-to-private/azure-cluster-public-to-private-commands.sh:
```diff
@@ -0,0 +1,212 @@
+    eval "${CMD}"
+}
+
+function check_image_registry_co() {
```
Maybe we can use `oc wait --for=condition=Ready co/image-registry --timeout=1200s` instead of this function.

jinyunma commented:
/pj-rehearse periodic-ci-openshift-verification-tests-master-installer-rehearse-4.18-installer-rehearse-azure

jianlinliu commented on ci-operator/step-registry/azure/provision/bastionhost/azure-provision-bastionhost-commands.sh:
```diff
@@ -181,20 +189,21 @@ vhd_blob_url=$(az storage blob url --account-name ${sa_name} --account-key ${acc
-    vnet_subnet_address_parameter="--address-prefixes ${bastion_subnet_cidr}"
+if [[ -z "${bastion_subnet}" ]]; then
+    echo "Create bastion subnet"
+    open_port="22 873 3128 3129 5000 6001 6002" bastion_nsg="${bastion_name}-nsg" bastion_subnet="${bastion_name}Subnet"
```
When using master subnet provided by classic IPI install, do we need to open these port in the nsg ?

jinyunma replied:
Fixed.

jianlinliu commented on ci-operator/step-registry/azure/cluster/public-to-private/azure-cluster-public-to-private-commands.sh:
```diff
@@ -0,0 +1,180 @@
+image_registry_spec_file=$(mktemp)
+oc get config.image/cluster -ojson | jq -r '.spec.storage.azure' | tee -a ${image_registry_spec_file}
+
+echo "Check spec of image-registry config have correct setting..."
```
Can these check be applied to a fresh installed private cluster ?

jianlinliu replied:
If these check also can be applied to private cluster, maybe we can move them to `cucushift-installer-check-azure-private`.

jinyunma replied:
moved check to step `cucushift-installer-check-azure-internal-registry` .
Xiujuan told me that we can also configure internal registry in common IPI cluster on day2, maybe this step can be reused.

xiuwang replied:
Yes, we can reuse this step in common IPI private cluster on day2.

jianlinliu commented on ci-operator/step-registry/azure/cluster/public-to-private/azure-cluster-public-to-private-commands.sh:
```diff
@@ -0,0 +1,180 @@
+
+# Check image-registry config
+image_registry_sa_name=$(az storage account list -g ${RESOURCE_GROUP} --query '[].name' -otsv | grep imageregistry)
+image_registry_private_endpoint=$(az network private-endpoint list -g ${RESOURCE_GROUP} --query '[].name' -otsv)
```
I saw common IPI private cluster does not have the private endpoint, is it not empty here in public-to-private case?

jinyunma replied:
private endpoint is the newly created by image registry when registry type is changed from External to Internal. In common IPI private cluster, the default type of registry is External, no private endpoint to be created.

jinyunma commented:
/pj-rehearse periodic-ci-openshift-verification-tests-master-installer-rehearse-4.18-installer-rehearse-azure

jinyunma commented:
/pj-rehearse periodic-ci-openshift-verification-tests-master-installer-rehearse-4.18-installer-rehearse-azure

jinyunma commented:
/pj-rehearse periodic-ci-openshift-verification-tests-master-installer-rehearse-4.18-installer-rehearse-azure

jianlinliu commented on ci-operator/step-registry/cucushift/installer/check/azure/internal-registry/cucushift-installer-check-azure-internal-registry-commands.sh:
```diff
@@ -0,0 +1,108 @@
+    source "${SHARED_DIR}/proxy-conf.sh"
+else
+  echo "No proxy-conf found, exit now"
+  exit 1
```
I guess a common IPI install also can configure a internal registry, so if there is no `proxy-conf.sh`, we can continue, not have to exit.

jinyunma commented:
/pj-rehearse periodic-ci-openshift-verification-tests-master-installer-rehearse-4.18-installer-rehearse-azure

jianlinliu commented:
/lgtm

jinyunma commented:
[Job](https://qe-private-deck-ci.apps.ci.l2s4.p1.openshiftapps.com/view/gs/qe-private-deck/pr-logs/pull/openshift_release/59442/rehearse-59442-periodic-ci-openshift-verification-tests-master-installer-rehearse-4.18-installer-rehearse-azure/1864510070268628992) failed at step of `cucushift-installer-check-azure-internal-registry`, image registry config spec was not refreshed on that time (05:00), actually it was updated finally(05:02) based on below log in pod `cluster-image-registry-operator`.

```
2024-12-05T05:00:50.375929558Z I1205 05:00:50.375874       1 azure.go:1020] setting azure storage account tags
2024-12-05T05:02:48.808455905Z I1205 05:02:48.808400       1 azure.go:751] storage account "imageregistryciop16hd7rm" is now served by private endpoint "imageregistryciop162m7c4". public network access is disabled
2024-12-05T05:02:48.832690117Z I1205 05:02:48.832638       1 controller.go:294] object changed: *v1.Config, Name=cluster (metadata=false, spec=true): added:spec.storage.azure.networkAccess.internal.privateEndpointName="imageregistryciop162m7c4", added:spec.storage.azure.networkAccess.internal.subnetName="ci-op-16ck6q73-11f79-l79x9-master-subnet", added:spec.storage.azure.networkAccess.internal.vnetName="ci-op-16ck6q73-11f79-l79x9-vnet", added:status.storage.azure.networkAccess.internal.privateEndpointName="imageregistryciop162m7c4", added:status.storage.azure.networkAccess.internal.subnetName="ci-op-16ck6q73-11f79-l79x9-master-subnet", added:status.storage.azure.networkAccess.internal.vnetName="ci-op-16ck6q73-11f79-l79x9-vnet", changed:status.storage.azure.networkAccess.type={"External" -> "Internal"}
```
seems that it is not enough to check image-registry operator status in step `azure-cluster-public-to-private`, maybe need sleeping between image-registry patch and operator status check.

jinyunma commented:
/pj-rehearse periodic-ci-openshift-verification-tests-master-installer-rehearse-4.18-installer-rehearse-azure

jianlinliu commented:
>  image registry config spec was not refreshed on that time (05:00), actually it was updated finally(05:02) based on below log in pod cluster-image-registry-operator

Sounds like the [fix](https://github.com/openshift/release/compare/468ddc414236e08b5e65328fbc311ebc3ee4f9c6..bc43377904b69bea65ed5549a870e80faf052e1b) does not help a lot.
`sleep 120s` happened prior to `oc wait --for=condition=Available --for condition=Progressing=False --for condition=Degraded=False co/image-registry --timeout=1200s`, I guess the sleep will save 120s for the wait command.

I guess once the operator get ready after the above wait command, `oc wait --for=condition=Ready=True pods --all -n openshift-image-registry --timeout=300s` always passed because the pod is always ready when co get ready.

If that makes sense, maybe we can introduce some retries in `cucushift-installer-check-azure-internal-registry` step.

jinyunma commented:
> Sounds like the [fix](https://github.com/openshift/release/compare/468ddc414236e08b5e65328fbc311ebc3ee4f9c6..bc43377904b69bea65ed5549a870e80faf052e1b) does not help a lot. `sleep 120s` happened prior to `oc wait --for=condition=Available --for condition=Progressing=False --for condition=Degraded=False co/image-registry --timeout=1200s`, I guess the sleep will save 120s for the wait command.

`sleep 120s` just makes sure that co becomes in progressing status, then check co status.
 I updated again, run `oc wait --for condition=Progressing=True co/image-registry --timeout=120s` instead of `sleep 120s`. Let's see if this can help.

jinyunma commented:
/pj-rehearse periodic-ci-openshift-verification-tests-master-installer-rehearse-4.18-installer-rehearse-azure

jianlinliu commented:
> > Sounds like the [fix](https://github.com/openshift/release/compare/468ddc414236e08b5e65328fbc311ebc3ee4f9c6..bc43377904b69bea65ed5549a870e80faf052e1b) does not help a lot. `sleep 120s` happened prior to `oc wait --for=condition=Available --for condition=Progressing=False --for condition=Degraded=False co/image-registry --timeout=1200s`, I guess the sleep will save 120s for the wait command.
> 
> `sleep 120s` just makes sure that co becomes in progressing status, then check co status. I updated again, run `oc wait --for condition=Progressing=True co/image-registry --timeout=120s` instead of `sleep 120s`. Let's see if this can help.

Ah, the original issue `oc wait --for=condition=Available --for condition=Progressing=False --for condition=Degraded=False co/image-registry --timeout=1200s` is executed too early, the operator did not start to update yet, so the wait almost completed immediately.

jinyunma commented:
/pj-rehearse periodic-ci-openshift-verification-tests-master-installer-rehearse-4.17-installer-rehearse-azure

jinyunma commented:
/pj-rehearse periodic-ci-openshift-verification-tests-master-installer-rehearse-4.18-installer-rehearse-azure

jinyunma commented:
workflow `azure-ipi-sharednetwork-public-to-private` is passed in 4.17 [job](https://qe-private-deck-ci.apps.ci.l2s4.p1.openshiftapps.com/view/gs/qe-private-deck/pr-logs/pull/openshift_release/59442/rehearse-59442-periodic-ci-openshift-verification-tests-master-installer-rehearse-4.17-installer-rehearse-azure/1864654189150343168).

workflow `azure-ipi-public-to-private` failed again in 4.18 [job](https://qe-private-deck-ci.apps.ci.l2s4.p1.openshiftapps.com/view/gs/qe-private-deck/pr-logs/pull/openshift_release/59442/rehearse-59442-periodic-ci-openshift-verification-tests-master-installer-rehearse-4.18-installer-rehearse-azure/1864611929658494976) on step `ipi-deprovision-deprovision`, this step export proxy before running command `openshift-install destroy cluster`. In failed workflow, bastion host is provisioned in cluster resource group, once bastion host is deleted during destroying, it causes other resources removing as proxy connection timeout. Try to [update step](https://github.com/openshift/release/compare/331b7c30e4c98351493f4e06faf58f2141310bb7..205f8be60333af0c9247b377c8178e6ca32ee9c0) `ipi-deprovision-deprovision` to skip exporting proxy for this job.

For panic error thrown by installer in `ipi-deprovision-deprovision`, not sure it caused by prow timeout or proxy connection timeout, will check it locally.

jianlinliu commented on ci-operator/step-registry/ipi/deprovision/deprovision/ipi-deprovision-deprovision-commands.sh:
```diff
@@ -101,6 +101,9 @@ fi
 if test -f "${SHARED_DIR}/proxy-conf.sh"; then
   if [[ "${CLUSTER_TYPE}" =~ ^aws-s?c2s$ ]]; then
     echo "proxy-conf.sh detected, but not reqquired by C2S/SC2S while destroying cluster, skip proxy setting"
+  elif [[ "${JOB_NAME}" =~ .*azure-ipi-public-to-private.* ]]; then
```
I guess we can use `CLUSTER_TYPE` for the case to avoid the job name hardcode.
```suggestion
  elif [[ "${CLUSTER_TYPE}" == "azure4" ]]; then
      # when bastion host is provisioned in cluster resource group, once the bastion is destroyed in the destroy process, the running destroy process would be interrupted and failed. E.g: azure-ipi-public-to-private jobs.
    echo "proxy-conf.sh detected, but not required by azure4 clusters while destroying cluster, skip proxy setting"
```

gpei commented:
looks like the job name in rehearsal is `periodic-ci-openshift-verification-tests-master-installer-rehearse-4.18-installer-rehearse-azure` so proxy is still being used in the deprovision.

jinyunma commented:
/pj-rehearse periodic-ci-openshift-verification-tests-master-installer-rehearse-4.18-installer-rehearse-azure

jianlinliu commented:
/lgtm

jinyunma commented:
/pj-rehearse ack

jinyunma commented:
cc @liangxia to review

liangxia commented:
/approve

liangxia commented:
BTW, we have 4.19 tests now. Not sure if we need to test this profile in 4.19, too ?

jinyunma commented:
> BTW, we have 4.19 tests now. Not sure if we need to test this profile in 4.19, too ?

Not noticed this, thanks for the info, I will add 4.19 job separately in new PR.
